torch/_subclasses/fake_tensor.py

```diff
@@ -1126,43 +1163,81 @@ def _validate_cache_key(
         convert FakeTensors into metadata. Raises _BypassDispatchCache to signal
         unsupported cases that should bypass caching.
         """
         if isinstance(args, dict):
-            args = list(args.keys()) + list(args.values())
+            self._prep_args_for_hash(output, args.keys())
+            self._prep_args_for_hash(output, args.values())
+            return
 
-        result: List[Any] = []
         for arg in args:
-            if isinstance(arg, FakeTensor):
-                if not self.is_our_fake(arg):
-                    raise _BypassDispatchCache("not our fake")
-                if arg._has_symbolic_sizes_strides:
-                    raise _BypassDispatchCache("symbolic shape")
-                if arg.constant is not None:
-                    raise _BypassDispatchCache("constant attribute")
-                if arg.is_sparse:
-                    raise _BypassDispatchCache("sparse tensor")
-                if is_sparse_compressed(arg):
-                    raise _BypassDispatchCache("sparse compressed tensor")
-                result.append(extract_tensor_metadata(arg))
-            elif isinstance(arg, torch.Tensor):
-                raise _BypassDispatchCache("non-fake tensor")
+            if isinstance(arg, torch.Tensor):
+                if isinstance(arg, FakeTensor):
+                    if arg._has_symbolic_sizes_strides:
+                        # This will get caught by _verify_args_for_hash later. We
+                        # can't just ignore it because it's an unhashable value. Use
+                        # a sentinel to indicate the unhashable value (so we don't
+                        # collide with a "good" hash).
+                        output.append(_UNHASHABLE)
+                    elif arg.fake_mode is not self:
+                        output.append(_UNHASHABLE)
+                    else:
+                        output.append(extract_tensor_metadata(arg))
+                else:
+                    # Caught by _verify_args_for_hash later.
+                    output.append(_UNHASHABLE)
             elif isinstance(arg, (torch.SymBool, torch.SymInt, torch.SymFloat)):
-                raise _BypassDispatchCache("symbolic shape")
+                # Caught by _verify_args_for_hash later.
+                output.append(_UNHASHABLE)
             elif isinstance(arg, (list, tuple, dict)):
-                result.extend(self._prep_args_for_hash(arg))
+                self._prep_args_for_hash(output, arg)
             else:
                 # It's important to capture the type of the arg since, e.g., 1 and 1.0
                 # hash to the same value, but can produce different dtypes for the
                 # output tensor.
-                result.append((type(arg), arg))
+                output.append(type(arg))
+                output.append(arg)
 
-        return tuple(result)
+    def _verify_args_for_hash(self, args: Any):
```

Comment:
If I'm reading this correctly, I think _prep_args_for_hash and _verify_args_for_hash need to stay in sync such that any time the former would append UNHASHABLE, the latter should raise _BypassDispatchCache. Is that right? I'm wondering if we could just note that we found an unhashable object during key calculation so that we don't have to re-check in `_validate_cache_key`?

Reply:
It shouldn't be a perf issue - we'll only do the `_validate_cache_key` check once. After that the failure will be cached.

Reply:
My question was more wondering if we could avoid having to keep these two sections of code in sync in case we add more conditions in the future.